lib/pac55xx/can.c

```diff
@@ -0,0 +1,460 @@
+			const uint8_t *data) {
+	/* if TBS is 0, then not ready to transmit */
+	if ((CAN_ISR_SR_CMR_MR(canport) & CAN_SR_TBS) == 0) {
+		return false;
+	}
+	uint32_t word = (length & CAN_BITS_3_0)
+		| (rtr ? BIT6 : 0) | BIT7 /* DLC/RTR/FF ==> 7:0 */
+		| ((id & CAN_BITS_28_21) >> 13) /* ID 28:21 ==> 15:8 */
+		| ((id & CAN_BITS_20_13) << 3)  /* ID 20:13 ==> 23:16 */
+		| ((id & CAN_BITS_12_5) << 19); /* ID 12:5 ==> 31:24 */
+	CAN_TXBUF(canport) = word; /* write first 32-bit word to FIFO */
+
+	word = ((id & CAN_BITS_4_0) << 3); /* ID 4:0 ==> 7:3 */
+	if (length > 0) {
+		word |= (data[0] << 8) | (data[1] << 16) | (data[2] << 24);
+	}
+	/* for extended frame, always write second 32-bit word to FIFO */
+	CAN_TXBUF(canport) = word;
+	if (length > 3) {
+		word = (data[3] << 0) | (data[4] << 8)
+			| (data[5] << 16) | (data[6] << 24);
+		CAN_TXBUF(canport) = word;
+	}
+	if (length > 7) {
+		word = data[7];
+		CAN_TXBUF(canport) = word;
+	}
+	/* Request transmit */
+	CAN_ISR_SR_CMR_MR_SET(canport, CAN_CMR_TR);
+	return true;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Abort Transmit
+Aborts the current transmission.
+
+@param[in] canport Unsigned int32. CAN block register base address.
+*/
+void can_abort_transmit(uint32_t canport) {
+	CAN_ISR_SR_CMR_MR_SET(canport, CAN_CMR_AT);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Receive Message
+If no data is in the RX buffer, id and length are set to 0.
+
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[out] id Unsigned int32 pointer. Message ID.
+@param[out] ext bool pointer. The message ID is extended.
+@param[out] rtr bool pointer. Remote Request bit value.
+@param[out] length Unsigned int8 pointer. Length of message payload.
+@param[out] data Unsigned int8[]. Message payload data, min length 8.
+*/
+void can_receive(uint32_t canport, uint32_t *id, bool *ext, bool *rtr, uint8_t *length,
+			uint8_t *data) {
+	if ((CAN_ISR_SR_CMR_MR(canport) & CAN_ISR_RI) == 0 || CAN_RMC(canport) == 0) {
+		*id = 0;
+		*length = 0;
+		return; /* empty RX FIFO */
+	}
+	uint32_t can_buffer = CAN_RXBUF(canport); /* read 32-bit word */
+	uint8_t rx_length = can_buffer & CAN_BITS_3_0;
+	bool is_extended = can_buffer & BIT7;
+	if (ext) {
+		*ext = is_extended;
+	}
+	if (rtr) {
+		*rtr = can_buffer & BIT6;
+	}
+	if (length) {
+		*length = rx_length;
+	}
+	uint32_t _id;
+	if (is_extended) {
+		/* Parse extended message ID from RXBUF */
+		_id = ((can_buffer & CAN_BITS_15_8) << 13)   /* ID 28:21 <== 15:8 */
+			| ((can_buffer & CAN_BITS_23_16) >> 3)   /* ID 20:13 <== 23:16 */
+			| ((can_buffer & CAN_BITS_31_24) >> 19); /* ID 12:5 <== 31:24 */
+		can_buffer = CAN_RXBUF(canport);
+		_id |= ((can_buffer & CAN_BITS_7_3) >> 3); /* ID 4:0 <== 7:3 */
+
+		/* Parse extended message data from RXBUF */
+		data[0] = can_buffer >> 8;
+		data[1] = can_buffer >> 16;
+		data[2] = can_buffer >> 24;
```

Comment:
shouldn't you be respecting rx_length here? are you just trusting that you're allowed to write whatever you like to the dest buffer?

Reply:
I added to the parameter doc that it must be 'min length 8'